File: 2/main.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    
    def add_helper(self, l1, l2, carry):
        if carry:
            if l1 is None and l2 is None:
                return ListNode(carry)
            elif l1 is None:
                digit = l2.val + carry
                ret = ListNode(digit % 10)
                ret.next = self.add_helper(l1, l2.next, digit / 10)
                return ret
            elif l2 is None:
                digit = l1.val + carry
                ret = ListNode(digit % 10)
                ret.next = self.add_helper(l1.next, l2, digit / 10)
                return ret
            else:
                digit = l1.val + l2.val + carry
                ret = ListNode(digit % 10)
                ret.next = self.add_helper(l1.next, l2.next, digit / 10)
                return ret
        else:
            if l1 is None and l2 is None:
                return None
            elif l1 is None:
                return l2
            elif l2 is None:
                return l1
            else:
                digit = l1.val + l2.val
                ret = ListNode(digit % 10)
                ret.next = self.add_helper(l1.next, l2.next, digit / 10)
                return ret
                

    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        return self.add_helper(l1, l2, 0)
        

File: 2/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:

    ListNode* add_helper(ListNode* l1, ListNode* l2, int carry) {
        if (carry) {
            if (l1 == NULL && l2 == NULL) {
                return new ListNode(carry);
            }
            else if (l1 == NULL) {
                int digit = l2->val + carry;
                ListNode *list = new ListNode(digit % 10);
                list->next = add_helper(l1, l2->next, digit / 10);
                return list;
            }
            else if (l2 == NULL) {
                int digit = l1->val + carry;
                ListNode *list = new ListNode(digit % 10);
                list->next = add_helper(l1->next, l2, digit / 10);
                return list;
            }
            else {
                int digit = l1->val + l2->val + carry;
                ListNode *list = new ListNode(digit % 10);
                list->next = add_helper(l1->next, l2->next, digit / 10);
                return list;
            }
        }
        else {
            if (l1 == NULL && l2 == NULL) {
                return NULL;
            }
            else if (l1 == NULL) {
                return l2;
            }
            else if (l2 == NULL) {
                return l1;
            }
            else {
                int digit = l1->val + l2->val;
                ListNode *list = new ListNode(digit % 10);
                list->next = add_helper(l1->next, l2->next, digit / 10);
                return list;
            }
        }
    }

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        return add_helper(l1, l2, 0);
    }
};